### PR TITLE
Allow Certificate names that aren't DNS-1035 labels.

### PR DESCRIPTION
**Pull request by tcnghia**
# Changes

- :bug: Allow Certificate names that aren't DNS-1035 labels. This fix is safe to roll out since previously we couldn't create the child resources (Service, Endpoints) using non DNS1035 anyway.

/kind bug

**Release Note**

```release-note
Make sure we can handle all kind of Certificate names.
```

### CODE REVIEW

File: pkg/reconciler/certificate/resources/service.go [hunk 1]
```diff
@@ -25,18 +25,29 @@ import (
...
 const portName = "http-challenge"
 
+func ServiceName(cert *v1alpha1.Certificate) string {
+	// Service names must be a DNS-1035 label. We try to use the
+	// Certificate name first if possible.
+	if name := kmeta.ChildName(cert.Name, ""); 0 == len(validation.IsDNS1035Label(name)) {
+		return name
+	}
+	// Fall back to a less readable name but guaranteed to be DNS-1035 label.
+	return "challenge-for-" + string(cert.GetUID())
```
Comment (mattmoor) [inline pkg/reconciler/certificate/resources/service.go:40]: Are we guaranteed that this fits or should we use `kmeta.ChildName()` to be sure?

Ack the fixed length, but your tests aren't using a full uuid, so I wanted to double check

Comment (tcnghia) [reply pkg/reconciler/certificate/resources/service.go]: Done now

File: pkg/reconciler/certificate/resources/service.go [hunk 2]
```diff
@@ -25,18 +25,32 @@ import (
...
+	// Service names must be a DNS-1035 label. We try to use the
+	// Certificate name first if possible.
+	if name := kmeta.ChildName(cert.Name, ""); 0 == len(validation.IsDNS1035Label(name)) {
+		return name
+	}
+	// Fall back to a less readable name but guaranteed to be DNS-1035 label.
+	//
+	// UUID are 35 chars long so we are under the 63 chars limit here,
+	// but we can use kmeta.ChildName here to be future-proof.
+	return kmeta.ChildName("challenge-for-"+string(cert.GetUID()), "")
```
Comment (mattmoor) [inline pkg/reconciler/certificate/resources/service.go:43]:
```suggestion
	return kmeta.ChildName("challenge-for-", string(cert.GetUID()))
```

**Review by mattmoor — COMMENTED** [review]
/ok-to-test
/lgtm
/approve

**Pull request merged**